Improve documentation about Kubernetes objects & managing them

sftim opened the pull request:
- ~Reorganize [Working with Kubernetes Objects](https://kubernetes.io/docs/concepts/overview/working-with-objects/) section [[preview](https://deploy-preview-31261--kubernetes-io-main-staging.netlify.app/docs/concepts/overview/working-with-objects/)]~
  - ~move [Understanding Kubernetes Objects](https://kubernetes.io/docs/concepts/overview/working-with-objects/kubernetes-objects/) to be section overview~
  - ~other tidying~
  - ~Signpost readers to more pages about Kubernetes objects~ _Now part of PR #42117_

---

- Improve docs relating to patching [[preview](https://deploy-preview-31261--kubernetes-io-main-staging.netlify.app/docs/reference/using-api/api-concepts/#patch-and-apply)] and SSA [[preview](https://deploy-preview-31261--kubernetes-io-main-staging.netlify.app/docs/reference/using-api/server-side-apply/)]

This will need tech review by someone familiar with how Kubernetes implements the `PATCH` HTTP verb. I wasn't able to follow the existing documentation well enough to be confident that this revised text is accurate.

/language en
/wg api-expression

sftim commented:
Helps with #31252

sftim commented on content/en/docs/reference/using-api/server-side-apply.md:
````diff
@@ -462,49 +505,51 @@ manager for kubectl server-side apply is `kubectl`.
 kubectl apply --server-side --field-manager=my-manager [--dry-run=server]
 ```
 
-## API Endpoint
-
-With the Server Side Apply feature enabled, the `PATCH` endpoint accepts the
-additional `application/apply-patch+yaml` content type. Users of Server Side
-Apply can send partially specified objects as YAML to this endpoint.  When
-applying a configuration, one should always include all the fields that they
-have an opinion about.
````
I removed this because:
- Server-Side Apply is GA and the feature is always available
- `PATCH` is a verb, not an endpoint
- the new Serialization section of the page covers the detail about the media type and format.

jimangel commented:
/lgtm

apelisse commented on content/en/docs/reference/using-api/api-concepts.md:
```diff
@@ -753,15 +754,59 @@ rules:
+  then the new one is saved.  
+  To Kubernetes, this is a **patch** operation.
+
+`application/strategic-merge-patch+json`
```
Note that this is not available to CRs, only built-in types.

sftim replied:
Good to know. BTW is that due to a bug, a planned feature, something else?

apelisse replied:
I don't know the exact history behind that, but we basically never found the time/envy to make it work for CRs, and then server-side apply started so it felt like a waste of time.

apelisse replied:
Also, schemas wasn't readily available initially to make these things work, so the work was never started.

liggitt replied:
strategic-merge-patch relies on complex perfectly-matching logic existing both in the client and the server (in the client to compute the patch, and in the server to apply the patch). in practice, this made it exceptionally hard (or impossible) to fix bugs in strategic merge patch application or use without breaking old clients against new servers, or new clients against old servers (xref https://github.com/kubernetes/kubernetes/pull/107549#discussion_r785029380)

We didn't really want to import those problems to custom resources.

apelisse commented on content/en/docs/reference/using-api/api-concepts.md:
```diff
@@ -753,15 +754,59 @@ rules:
 client-side functionality of `kubectl apply`.
 
-The API verb for Server-Side Apply is **apply**.
+The API verb for Server-Side Apply is **apply**, whereas the API verb for other
```
I'm not exactly sure what this is supposed to mean?

sftim replied:
If I want to allow someone to make a change to an object using Server-Side Apply, what does the RBAC grant look like? Do I grant `apply`, or `patch`, or `update`?

liggitt replied:
`patch`

liggitt replied:
(if the object does not currently exist and server-side-apply would create the object, they need `create` permission as well)

apelisse commented on content/en/docs/reference/using-api/server-side-apply.md:
```diff
@@ -7,73 +7,86 @@ reviewers:
 {{< feature-state for_k8s_version="v1.22" state="stable" >}}
 
-## Introduction
+Kubernetes supports multiple appliers collaborating to manage the `.spec` and / or
```
I don't know if that's correct, we also allow multiple appliers on `.status`

sftim replied:
I think that's existing wording from somewhere. What's the true statement?

apelisse replied:
You can change all the same things with apply that you can change with a regular PUT/PATCH

sftim replied:
:thinking: If two controllers / API clients are collaborating on managing `.status`, they might also need write access to the resource as well as to its `status` subresource?

The important detail here is that SSA supports the “multiple co-operating clients” / blackboard pattern. I'll try to reword to avoid implying that SSA can't manage `.status`.

apelisse commented on content/en/docs/reference/using-api/server-side-apply.md:
```diff
@@ -7,73 +7,86 @@ reviewers:
-field.  If the field is not owned by any other field managers, it is either
-deleted from the live object or reset to its default value, if it has one. The
-same rule applies to associative list or map items.
+If you have Server-Side Apply enabled, the Kubernetes control plane tracks
```
That's true by default, possibly the option of disabling will be removed, should we make it clearer that this is the default?

sftim replied:
I can add that it's enabled by default.

sftim commented:
I've also asked WG API Expression for a tech review.

apelisse reviewed (commented):
Thanks for doing this Tim!

sftim commented:
Needs some work
/lgtm cancel

sftim commented:
How's this looking now?

apelisse reviewed (commented):
minor nits, thanks!

kbhawkey commented on content/en/docs/concepts/overview/working-with-objects/_index.md:
```diff
@@ -1,7 +1,110 @@
+The precise format of the object `spec` is different for every Kubernetes object, and contains nested fields specific to that object. The [Kubernetes API Reference](https://kubernetes.io/docs/reference/kubernetes-api/) can help you find the spec format for all of the objects you can create using Kubernetes.
+
+For example, the reference for Pod details the [`spec` field](/docs/reference/kubernetes-api/workload-resources/pod-v1/#PodSpec)
+for a Pod in the API, and the reference for Deployment details the [`spec` field](/docs/reference/kubernetes-api/workload-resources/deployment-v1/#DeploymentSpec) for Deployments.
```
Lines 88-90, do these sentences add significant value?

sftim replied:
The rebase should update these.

kbhawkey commented:
@sftim, Do you need to add a redirect for, `content/en/docs/concepts/overview/working-with-objects/kubernetes-objects.md`?

kbhawkey commented:
Should the "What is Kubernetes" section list before "Objects in Kubernetes" in Overview?
https://deploy-preview-31261--kubernetes-io-main-staging.netlify.app/docs/concepts/overview/

To review:
https://deploy-preview-31261--kubernetes-io-main-staging.netlify.app/docs/reference/using-api/server-side-apply/

Content moved to index page (with section links)
https://deploy-preview-31261--kubernetes-io-main-staging.netlify.app/docs/concepts/overview/working-with-objects/

apelisse commented on content/en/docs/concepts/overview/working-with-objects/_index.md:
```diff
@@ -1,7 +1,110 @@
+* The resources available to those applications
+* The policies around how those applications behave, such as restart policies, upgrades, and fault-tolerance
+
+A Kubernetes object is a "record of intent"—once you create the object, the Kubernetes system will constantly work to ensure that object exists. By creating an object, you're effectively telling the Kubernetes system what you want your cluster's workload to look like; this is your cluster's *desired state*.
```
Absolutely, I'm fine with that, thanks.

kbhawkey commented on content/en/docs/reference/using-api/api-concepts.md:
```diff
@@ -753,15 +754,63 @@ rules:
 See [Authorization Overview](/docs/reference/access-authn-authz/authorization/).
 
-## Server Side Apply
+## Updating existing resources {#patch-and-apply}
```
nit: Update resources

kbhawkey commented on content/en/docs/reference/using-api/server-side-apply.md:
```diff
@@ -292,82 +351,75 @@ Before `spec.data` gets changed from `atomic` to `granular`,
 
 In addition to the concurrency controls provided by [conflict resolution](#conflicts),
-Server Side Apply provides ways to perform coordinated
+Server-Side Apply provides ways to perform coordinated
```
The feature is called "Server Side Apply"?

sftim replied:
Tech reviewers: what is the right name?

apelisse replied:
It's fairly inconsistent, but I would say that `Server-Side Apply` wins.

```sh
apelisse ~/code/website> rg -i "server side apply" -c
content/zh/docs/reference/using-api/api-concepts.md:4
content/zh/docs/reference/using-api/server-side-apply.md:18
content/en/docs/reference/using-api/api-concepts.md:4
content/en/docs/reference/using-api/server-side-apply.md:18
apelisse ~/code/website> rg -i "server-side apply" -c
content/zh/blog/_posts/2020-03-25-kubernetes-1.18-release-announcement.md:1
content/en/blog/_posts/2020-04-01-server-side-apply-beta2.md:7
content/en/blog/_posts/2018-07-16-kubernetes-1-11-release-interview.md:1
content/en/blog/_posts/2019-01-14-apiserver-dry-run-and-kubectl-diff.md:2
content/en/blog/_posts/2020-03-25-kubernetes-1.18-release-announcement.md:1
content/en/blog/_posts/2019-06-20-Future-of-CRDs-Structural-Schemas.md:1
content/ko/docs/setup/release/notes.md:2
content/zh/docs/tasks/manage-kubernetes-objects/declarative-config.md:1
content/zh/docs/setup/release/notes.md:1
content/zh/docs/reference/using-api/server-side-apply.md:15
content/en/docs/tasks/manage-kubernetes-objects/declarative-config.md:1
content/en/docs/reference/using-api/server-side-apply.md:13
static/docs/reference/generated/kubectl/kubectl-commands.html:2
```

kbhawkey commented on content/en/docs/reference/using-api/server-side-apply.md:
```diff
@@ -198,81 +256,82 @@ data:
   name: foo-sample
   managedFields:
-  - manager: manager-one
+  - manager: "manager-one"
```
nit: quotes?

apelisse replied:
I don't personally have an opinion on that.

sftim replied:
I added the quotes to make it a bit more clear that a manager name is an opaque (?) string. Even though YAML lets you leave it out.

kbhawkey commented on content/en/docs/reference/using-api/server-side-apply.md:
```diff
@@ -198,81 +256,82 @@ data:
-structs. These markers can be applied to objects of the respective type,
-in Go files or in the OpenAPI schema definition of the
-[CRD](/docs/reference/generated/kubernetes-api/{{< param "version" >}}#jsonschemaprops-v1-apiextensions-k8s-io):
+The Kubernetes API (and the Go code that implementes that APU for Kubernetes) allows
```
nit: What is APU?

kbhawkey commented:
Ping for review
/wg api-expression

kbhawkey commented:
👋 @sftim . What do you think about moving forward with these changes?

sftim commented:
Edits since [informal LGTM](https://github.com/kubernetes/website/pull/31261#pullrequestreview-901957299) are minor. I also rebased against _main_.

tengqm commented on content/en/docs/concepts/overview/working-with-objects/_index.md:
```diff
@@ -1,7 +1,150 @@
 ---
-title: "Working with Kubernetes Objects"
-weight: 40
+title: Objects in Kubernetes
```
What is the "record of intent" for moving the text from "working-with-objects/kubernetes-objects.md" to "working-with-objects/_index.md" ?

A side effect I can perceive is the new redirect entry which implies an additional round trip between my browser and the web server.

sftim replied:
Is https://github.com/kubernetes/website/pull/31261/commits/997db9a88d7caabf4be0dccc84a63f4de8058060 the answer you were looking for @tengqm ?

I was worried that https://kubernetes.io/docs/concepts/overview/working-with-objects/ isn't very approachable for a newcomer, so I opted to move the overview page one level up.

I could add an extra commit to find pages that link to the old URI and change them to reference the new one.  Should I do that?

tengqm replied:
I think the existing page is very approachable. We have a shortcut on the "front page".

<img width="827" alt="image" src="https://user-images.githubusercontent.com/5094756/183394846-639a8ede-3b69-45e0-9424-12cd2b45b56b.png">

sftim replied:
I was imagining that some readers might arrive at https://kubernetes.io/docs/concepts/overview/working-with-objects/ by working through the left-hand nav of the site structure, with each page shown in order.

tengqm replied:
Right. If we want to improve the index page, why don't we just replace the generated indice by a tweaked list of links. If we change this `_index.md` file as proposed in this PR, users will likely to have difficulties finding the entries to topics like "name and IDs", "annotations", "field and selectors" etc. Those pages are deeply nested already and they are not shown on the left navigation.

If you want to improve the object concept page, please focus you change to the `working_with_objects/kubernets_objects.md` page rather than moving it elsewhere. If you can convince me that this relocation is a good move, please propose it in a separate PR.

/hold

tengqm commented on content/en/docs/reference/using-api/api-concepts.md:
```diff
@@ -753,15 +754,68 @@ rules:
+
+Instead of sending a PUT request, the client can send an instruction to the API
+server to **patch** an existing resource. A **patch** is appropriate if the
+change that the client wants to make isn't conditional on the existing data.
```
I don't think this (conditional or not) is the top of line consideration when requesting a patch.
A patch operation is mostly about "partial update", which means I don't want to **replace** an existing object. I'm interested in changing one or more attributes as the desired state.
Considering the "level-triggering" nature of k8s API, most of the time the server doesn't care about "lost updates". This is especially true for cases like scaling a replica set.

sftim replied:
> Considering the "level-triggering" nature of k8s API, most of the time the server doesn't care about "lost updates". This is especially true for cases like scaling a replica set.

I've definitely seen people suggest performing patches that could lose updates. Incrementing the scale of a ReplicaSet or Deployment? Probably fine. Updating a specific key within a ConfigMap that represents a _last modified_ timestamp? You can do it, but it's up the client to pick the right operation.

tengqm commented on content/en/docs/reference/using-api/server-side-apply.md:
```diff
@@ -7,75 +7,82 @@ reviewers:
 - liggitt
 content_type: concept
```
Shall we separate server-side-apply changes into another PR?
We are mixing a lot of changes in this one.

sftim replied:
This is now mostly about SSA and patching.

apelisse commented on content/en/docs/reference/using-api/api-concepts.md:
```diff
@@ -753,15 +754,68 @@ rules:
+that `resourceVersion` information so that the API server can detect lost updates
+and reject requests made by a client that is out of date with the cluster.
+In the event that the resource has changed (the `resourceVersion` the clienti
+provided is stale), the API server returns a `409 Conflict` error response.
```
I don't know if that whole paragraph properly serves its purpose. This is basically introducing [optimistic locking](https://en.wikipedia.org/wiki/Optimistic_concurrency_control), and I don't know if that description makes that clear. I would also probably mention that PUT must almost exclusively be used in a "read-modify-write" pattern which the "taking this from the object being updated" makes confusing to some extent.

Also note that, as mentioned in the Wikipedia page for optimistic locking, that mostly works for low data contention systems (and patch on the other hand helps avoid that problem, see [discussion below](https://github.com/kubernetes/website/pull/31261/files#r939597111)).

apelisse replied:
```suggestion
provided is stale), the API server returns a `409 Conflict` error response.

A common use-case is called "read-modify-write", one would first GET the object,
unmarshal it, modify it, re-marshal it before doing a PUT to write it back. That doesn't
require to worry about that `resourceVersion` that will get carried from the GET automatically.
```

Maybe, or something like this. I think the phrasing of the first paragraph is correct, but confusing if you don't know how to use it properly, which that second paragraph is trying to address.

apelisse commented on content/en/docs/reference/using-api/api-concepts.md:
```diff
@@ -753,15 +754,68 @@ rules:
+server to **patch** an existing resource. A **patch** is appropriate if the
+change that the client wants to make isn't conditional on the existing data.
+Clients that need effective detection of lost updates should consider using
+HTTP PUT, and retrying in case of conflict.
```
Two additional things:
1. You can have conditions in your json-patch to validate some things are consistent when the patch is applied (or applying will fail with a 40x bad request)
2. You can also enforce a specific resource version in the patch, which will give you the PUT guarantees without having to send the entire object.

sftim replied:
@apelisse Do we already have existing docs that show how to set a condition for a patch?

apelisse replied:
I don't think we have one. I know this because I tried and also now because I know how this works internally. That's partly why I wanted to mention it here :-)

sftim replied:
I reworded, mentioning that (in outline) under `application/json-patch+json`.

apelisse commented on content/en/docs/reference/using-api/api-concepts.md:
```diff
@@ -753,15 +754,68 @@ rules:
+  or a **patch** operation if the object already exists.
+
+`application/json-patch+json`
+: JSON Patch, as defined in [RFC6902](https://tools.ietf.org/html/rfc6902).
```
I don't know if it's worth mentioning that one can use the "test" operation of json-patch to guarantee some consistency in their patch.

sftim replied:
Could that go in a separate PR?

“guarantee” and “some” don't sit well with each other. What's the story here?

sftim replied:
Done (in outline). 

[I hope that] details can come in a follow-up PR by someone who understands it.

apelisse replied:
Looks good to me thanks

divya-mohan0209 commented on content/en/docs/concepts/overview/working-with-objects/_index.md:
```diff
@@ -1,7 +1,150 @@
+
+Almost every Kubernetes object includes two nested object fields that govern
+the object's configuration: the object *`spec`* and the object *`status`*.
+For objects that have a `spec`, you have to set this when you create the object,
```
Would this make more sense since "this" sounds very vague?
```suggestion
For objects that have a `spec`, you have to set it when you create the object,
```

divya-mohan0209 commented on content/en/docs/reference/using-api/api-concepts.md:
```diff
@@ -753,15 +754,68 @@ rules:
+the `resourceVersion` (taking this from the object being updated). Kubernetes uses
+that `resourceVersion` information so that the API server can detect lost updates
+and reject requests made by a client that is out of date with the cluster.
+In the event that the resource has changed (the `resourceVersion` the clienti
```
nit: Typo
```suggestion
In the event that the resource has changed (the `resourceVersion` the client
```

divya-mohan0209 reviewed (commented):
General nits.

tengqm commented on content/en/docs/concepts/overview/working-with-objects/_index.md:
```diff
@@ -1,7 +1,150 @@
+* The policies around how those applications behave, such as restart policies, upgrades, and
+  fault-tolerance
+
+A Kubernetes object is a "record of intent"—once you create the object, the Kubernetes system
```
```suggestion
A Kubernetes object is a "record of intent" — once you create the object, the Kubernetes system
```

sftim replied:
Not sure what our style guide says about em dashes. Using em dashes without surrounding space is usual for UK typography.

tengqm commented on content/en/docs/concepts/overview/working-with-objects/_index.md:
```diff
@@ -1,7 +1,150 @@
+telling the Kubernetes system what you want your cluster's workload to look like; this is your
+cluster's *desired state*.
+
+To work with Kubernetes objects—whether to create, modify, or delete them—you'll need to
```
```suggestion
To work with Kubernetes objects — whether to create, modify, or delete them — you'll need to
```

sftim commented:
I've paused this to focus on getting an even older PR (30817) merged in.

sftim commented:
If I split this into multiple PRs, I could also look at addressing https://github.com/kubernetes/website/issues/30576

divya-mohan0209 commented:
Hey @sftim , I know you're swamped and are working on getting #30817 ready. Are there any updates on this? Whenever you have some free cycles, please could you rebase?

sftim commented:
I had another go at this. It looks OK to me; I haven't looked thoroughly (sorry).

sftim commented:
@tengqm [requested](https://github.com/kubernetes/website/pull/31261#discussion_r973507041) that moving the overview page up one level happened in its own PR.

We've done that in a few places, I can make another change for that. I have not done this yet, so that feedback stands.

sftim commented:
I'm going to unhold this because https://github.com/kubernetes/website/pull/39576 has merged separately.

sftim commented:
/hold cancel

sftim commented:
I'd like another review on this. My guess is that the overall change is net beneficial enough to swing the balance in favor of a merge.

There may be a nit or two remaining.

sftim commented:
@jimangel if I could call in a favor, you did previously review and LGTM this change. Another look would be welcome.

apelisse commented:
I'll try to give it another look too, thanks for working on this @sftim!

sftim commented on content/en/docs/reference/using-api/server-side-apply.md:
```diff
@@ -213,67 +284,80 @@ more stable object lifecycle. Server-Side Apply tries to merge fields based on
 
-When a list, map, or struct changes from `atomic` to
-`map`/`set`/`granular`, the API server won't be able to infer the new
+When a `listType`, `mapType`, or `structType` changes from `atomic` to
```
Is this better? Happy to keep the original text here if not.

sftim commented on content/en/docs/reference/using-api/server-side-apply.md:
```diff
@@ -293,33 +377,22 @@ Before `spec.data` gets changed from `atomic` to `granular`,
 without a conflict), but it no longer owns `key1` and `key2`, so another
 manager can then modify or delete those fields without conflict.
 
-## Custom Resources
-
-By default, Server-Side Apply treats custom resources as unstructured data. All
-keys are treated the same as struct fields, and all lists are considered atomic.
-
-If the Custom Resource Definition defines a
-[schema](/docs/reference/generated/kubernetes-api/{{< param "version" >}}#jsonschemaprops-v1-apiextensions-k8s-io)
-that contains annotations as defined in the previous "Merge Strategy"
-section, these annotations will be used when merging objects of this
-type.
```
Moved; search for `## Custom resources and Server-Side Apply`

apelisse commented on content/en/docs/reference/using-api/api-concepts.md:
```diff
@@ -937,15 +938,74 @@ rules:
+In the event that the resource has changed (the `resourceVersion` the client
+provided is stale), the API server returns a `409 Conflict` error response.
+
+Instead of sending a PUT request, the client can send an instruction to the API
```
Unfortunately, we're not really addressing the question of "Why would I do one instead of the other?"

apelisse replied:
Making this a little longer than I would want but I'm being asked these questions ALL THE TIME, and I'm still looking for the perfect reference to point to, so let's take that chance:

UPDATE is great and easy, a couple of problems:
1. You might accidentally drop fields if you decode into Go objects that are not in the latest version (new fields are going to be dropped)
2. If there's a lot of contention on the object (even on a part that you're not trying to edit), you might have trouble sending the update. (you are sort of mentioning that in the next sentence but I don't know if that's super clear).
3. Related to 2, the object might be a lot bigger, which when there is contention, makes it even harder to update quickly.
4. You can easily depend on some values of the object

PATCH is great because:
1. You're sending only the piece that you want to change, the change is smaller
2. You don't depend on the rest of the object (as you mentioned in the next sentence)
3. Might require you to manually set the resourceVersion if you want to be super careful
5. you need to be careful when building the patch to not drop un-deserialized fields (order of the operation matters)
6. you need more local logic to build the patch (depending on language and presence of proper json patch libraries)
7. You can still fairly well depend on other values of the object
8. If you're using json-patch, you can use `test` conditions to careful craft specific update conditions

Apply is awesome because:
1. Doesn't require a GET before-hand
2. You still get conflicts if you change something that is unexpected, which significantly reduces the retry-on-error loop for conflicts, but they can easily be forced
3. Extremely easy to build
4. Does not work at all for values that depend on a current value of the object

sftim replied:
I added a new section, see fragment `#update-mechanism-choose` of the revised doc.

apelisse commented on content/en/docs/reference/using-api/server-side-apply.md:
```diff
@@ -7,75 +7,97 @@ reviewers:
 
-Field management is stored in a`managedFields` field that is part of an object's
-[`metadata`](/docs/reference/generated/kubernetes-api/{{< param "version" >}}/#objectmeta-v1-meta).
+If you remove a field from a manifest and apply that manifest, Server-Side
```
This sentence is not finished :-)

apelisse commented on content/en/docs/reference/using-api/server-side-apply.md:
```diff
@@ -394,41 +470,54 @@ If the apply results in a conflict with the HPA controller, then do nothing. The
-Users can transfer ownership of a field between each other by setting the field
-to the same value in both of their applied configs, causing them to share
-ownership of the field. Once the users share ownership of the field, one of them
+Field manages can transfer ownership of a field between each other by setting the field
```
```suggestion
Field managers can transfer ownership of a field between each other by setting the field
```

apelisse commented on content/en/docs/reference/using-api/server-side-apply.md:
```diff
@@ -477,48 +566,75 @@ manager for kubectl server-side apply is `kubectl`.
 first and the other changes being processed afterwards. As a result the
 applier takes ownership of any fields updated in the same request.
+
+{{< caution >}}
+Server Side Apply does not correctly track ownership on
+sub-resources that don't receive the resource object type. If you are
+using Server Side Apply with such a sub-resource, the changed fields
+won't be tracked.
+{{< /caution >}}
```
I don't know if that's entirely true, I think it should work for `/scale`.

apelisse reviewed (commented):
I'm happy to get this merged and iterate on top of this. This has been here for a long time and we would benefit from any change at that point. Thank you!

tengqm commented:
/lgtm

sftim commented:
Changes since last reviews:

- fixed typos etc
- added a new section about choosing which approach for updating existing objects
- added an entry for the client-side apply annotation, and linked to it

jimangel commented:
/lgtm 🚀

apelisse commented:
Thanks!

sftim commented:
Also see https://github.com/kubernetes/website/issues/42524

reylejano commented:
Applying lgtm from Jim's [comment](https://github.com/kubernetes/website/pull/31261#issuecomment-1656143667) 
/lgtm

reylejano commented:
The changes in the is PR look good and is an improvement
Thank you for addressing comments and questions
/label refactor
/approve
